File: core/search_query.py
"""
Advanced paper search: field filters, quoted phrases, negation, and OR branches.

Syntax (case-insensitive except quoted text is matched as given after lowercasing):
  • Free words — all must appear somewhere in title, content, folder, deck, or tags (AND).
  • "phrase" or 'phrase' — exact substring in that combined haystack.
  • title:text, content:text, folder:text, deck:text, tag:text — restrict to a field;
    use quotes for spaces: folder:"Cell Biology"
  • -word or -field:value — exclude papers that match.
  • OR — alternative branches:  a b OR c d  matches if (a AND b) OR (c AND d).
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Tuple

KNOWN_FIELDS = frozenset({"title", "content", "folder", "deck", "tag"})


@dataclass
class SearchBranch:
    """One AND-group (terms + field filters + negations)."""

    terms: List[str] = field(default_factory=list)
    fields: List[Tuple[str, str]] = field(default_factory=list)
    neg_terms: List[str] = field(default_factory=list)
    neg_fields: List[Tuple[str, str]] = field(default_factory=list)

    def matches(self, paper: Any) -> bool:
        h = _haystack_default(paper)
        for t in self.terms:
            if t not in h:
                return False
        for fname, needle in self.fields:
            if needle not in _get_field_text(paper, fname):
                return False
        for t in self.neg_terms:
            if t in h:
                return False
        for fname, needle in self.neg_fields:
            if needle in _get_field_text(paper, fname):
                return False
        return True


def _get_field_text(paper: Any, fname: str) -> str:
    if fname == "title":
        return (paper.title or "").lower()
    if fname == "content":
        return (paper.content or "").lower()
    if fname == "folder":
        return (paper.folder_path or "").lower()
    if fname == "deck":
        return (paper.deck_name or "").lower()
    if fname == "tag":
        tags = getattr(paper, "tags", None) or []
        return " ".join((t or "").lower() for t in tags)
    return ""


def _haystack_default(paper: Any) -> str:
    tags = getattr(paper, "tags", None) or []
    parts = [
        paper.title or "",
        paper.content or "",
        paper.folder_path or "",
        paper.deck_name or "",
        " ".join(str(t) for t in tags if t),
    ]
    return "\n".join(parts).lower()


def split_or_branches(q: str) -> List[str]:
    """Split on top-level ` OR ` (not inside quotes)."""
    q = q.strip()
    if not q:
        return []
    parts: List[str] = []
    buf: List[str] = []
    i = 0
    n = len(q)
    while i < n:
        ch = q[i]
        if ch in "\"'":
            quote = ch
            buf.append(ch)
            i += 1
            while i < n and q[i] != quote:
                buf.append(q[i])
                i += 1
            if i < n:
                buf.append(q[i])
                i += 1
            continue
        if (
            i + 3 < n
            and q[i : i + 4].upper() == " OR "
            and (i == 0 or q[i - 1].isspace())
        ):
            j = i + 4
            if j >= n or q[j].isspace():
                chunk = "".join(buf).strip()
                if chunk:
                    parts.append(chunk)
                buf = []
                i = j
                while i < n and q[i].isspace():
                    i += 1
                continue
        buf.append(ch)
        i += 1
    tail = "".join(buf).strip()
    if tail:
        parts.append(tail)
    return parts if parts else [q]


def parse_branch(s: str) -> SearchBranch:
    b = SearchBranch()
    i = 0
    n = len(s)

    def skip_ws() -> None:
        nonlocal i
        while i < n and s[i].isspace():
            i += 1

    while True:
        skip_ws()
        if i >= n:
            break
        neg = False
        if s[i] == "-":
            neg = True
            i += 1
            skip_ws()
        if i >= n:
            break
        m = re.match(r"(title|content|folder|deck|tag):", s[i:], re.I)
        if m:
            fname = m.group(1).lower()
            i += m.end()
            skip_ws()
            if i >= n:
                break
            if s[i] in "\"'":
                quote = s[i]
                i += 1
                start = i
                while i < n and s[i] != quote:
                    i += 1
                val = s[start:i].lower()
                if i < n:
                    i += 1
            else:
                start = i
                while i < n and not s[i].isspace():
                    i += 1
                val = s[start:i].lower()
            if fname in KNOWN_FIELDS and val != "":
                if neg:
                    b.neg_fields.append((fname, val))
                else:
                    b.fields.append((fname, val))
            continue
        if s[i] in "\"'":
            quote = s[i]
            i += 1
            start = i
            while i < n and s[i] != quote:
                i += 1
            val = s[start:i].lower()
            if i < n:
                i += 1
            if val:
                if neg:
                    b.neg_terms.append(val)
                else:
                    b.terms.append(val)
            continue
        start = i
        while i < n and not s[i].isspace():
            i += 1
        val = s[start:i].lower()
        if val:
            if neg:
                b.neg_terms.append(val)
            else:
                b.terms.append(val)
    return b


def parse_paper_search(query: str) -> List[SearchBranch]:
    return [parse_branch(part) for part in split_or_branches(query)]


def _snippet_around(content: str, idx: int, length: int, radius: int = 48) -> str:
    start = max(0, idx - radius)
    end = min(len(content), idx + length + radius)
    frag = content[start:end].replace("\n", " ")
    return ("..." if start > 0 else "") + frag + ("..." if end < len(content) else "")


def build_snippet(paper: Any, br: SearchBranch) -> str:
    content = paper.content or ""
    if not content:
        return ""
    cl = content.lower()
    for fname, needle in br.fields:
        if fname == "content" and needle and needle in cl:
            return _snippet_around(content, cl.index(needle), len(needle))
    for t in br.terms:
        if t and t in cl:
            return _snippet_around(content, cl.index(t), len(t))
    for fname, needle in br.fields:
        if fname != "content" and needle and needle in cl:
            return _snippet_around(content, cl.index(needle), len(needle))
    return ""


def match_flags(paper: Any, br: SearchBranch) -> Dict[str, bool]:
    title_l = (paper.title or "").lower()
    cl = (paper.content or "").lower()
    folder_l = (paper.folder_path or "").lower()
    deck_l = (paper.deck_name or "").lower()
    tags_l = " ".join((t or "").lower() for t in (getattr(paper, "tags", None) or []))

    title_match = any(t in title_l for t in br.terms) or any(
        f == "title" and n in title_l for f, n in br.fields
    )
    content_match = any(t in cl for t in br.terms) or any(
        f == "content" and n in cl for f, n in br.fields
    )
    folder_match = any(t in folder_l for t in br.terms) or any(
        f == "folder" and n in folder_l for f, n in br.fields
    )
    deck_match = any(t in deck_l for t in br.terms) or any(
        f == "deck" and n in deck_l for f, n in br.fields
    )
    tag_match = any(t in tags_l for t in br.terms) or any(
        f == "tag" and n in tags_l for f, n in br.fields
    )
    return {
        "title_match": title_match,
        "content_match": content_match,
        "folder_match": folder_match,
        "deck_match": deck_match,
        "tag_match": tag_match,
    }


def search_papers_advanced(papers: List[Any], query: str) -> List[Dict[str, Any]]:
    q = query.strip()
    if not q:
        return []
    branches = parse_paper_search(q)
    out: List[Dict[str, Any]] = []
    for p in papers:
        matched: Optional[SearchBranch] = None
        for br in branches:
            if br.matches(p):
                matched = br
                break
        if matched is None:
            continue
        flags = match_flags(p, matched)
        snippet = build_snippet(p, matched)
        out.append(
            {
                "id": p.id,
                "title": p.title,
                "folder_path": p.folder_path,
                "deck_name": getattr(p, "deck_name", "") or "",
                "snippet": snippet,
                **flags,
            }
        )
    return out

File: web_src/src/searchQuery.js
/**
 * Client-side mirror of core/search_query.py for mock bridge / dev without Anki.
 */

const KNOWN = new Set(['title', 'content', 'folder', 'deck', 'tag']);

function haystack(p) {
  const tags = p.tags || [];
  return [p.title, p.content, p.folder_path, p.deck_name, tags.join(' ')]
    .map((x) => String(x || '').toLowerCase())
    .join('\n');
}

function fieldText(p, fname) {
  if (fname === 'title') return String(p.title || '').toLowerCase();
  if (fname === 'content') return String(p.content || '').toLowerCase();
  if (fname === 'folder') return String(p.folder_path || '').toLowerCase();
  if (fname === 'deck') return String(p.deck_name || '').toLowerCase();
  if (fname === 'tag') return (p.tags || []).map((t) => String(t || '').toLowerCase()).join(' ');
  return '';
}

function splitOrBranches(q) {
  const s = q.trim();
  if (!s) return [];
  const parts = [];
  const buf = [];
  let i = 0;
  const n = s.length;
  while (i < n) {
    const ch = s[i];
    if (ch === '"' || ch === "'") {
      const quote = ch;
      buf.push(ch);
      i++;
      while (i < n && s[i] !== quote) {
        buf.push(s[i]);
        i++;
      }
      if (i < n) buf.push(s[i]);
      i++;
      continue;
    }
    if (
      i + 3 < n &&
      s.slice(i, i + 4).toUpperCase() === ' OR ' &&
      (i === 0 || /\s/.test(s[i - 1]))
    ) {
      const j = i + 4;
      if (j >= n || /\s/.test(s[j])) {
        const chunk = buf.join('').trim();
        if (chunk) parts.push(chunk);
        buf.length = 0;
        i = j;
        while (i < n && /\s/.test(s[i])) i++;
        continue;
      }
    }
    buf.push(ch);
    i++;
  }
  const tail = buf.join('').trim();
  if (tail) parts.push(tail);
  return parts.length ? parts : [s];
}

function parseBranch(s) {
  const branch = {
    terms: [],
    fields: [],
    negTerms: [],
    negFields: [],
  };
  let i = 0;
  const n = s.length;
  const skipWs = () => {
    while (i < n && /\s/.test(s[i])) i++;
  };
  const fieldRe = /^(title|content|folder|deck|tag):/i;

  while (true) {
    skipWs();
    if (i >= n) break;
    let neg = false;
    if (s[i] === '-') {
      neg = true;
      i++;
      skipWs();
    }
    if (i >= n) break;
    const rest = s.slice(i);
    const fm = rest.match(fieldRe);
    if (fm) {
      const fname = fm[1].toLowerCase();
      i += fm[0].length;
      skipWs();
      if (i >= n) break;
      let val;
      if (s[i] === '"' || s[i] === "'") {
        const quote = s[i];
        i++;
        const start = i;
        while (i < n && s[i] !== quote) i++;
        val = s.slice(start, i).toLowerCase();
        if (i < n) i++;
      } else {
        const start = i;
        while (i < n && !/\s/.test(s[i])) i++;
        val = s.slice(start, i).toLowerCase();
      }
      if (KNOWN.has(fname) && val !== '') {
        if (neg) branch.negFields.push([fname, val]);
        else branch.fields.push([fname, val]);
      }
      continue;
    }
    if (s[i] === '"' || s[i] === "'") {
      const quote = s[i];
      i++;
      const start = i;
      while (i < n && s[i] !== quote) i++;
      const val = s.slice(start, i).toLowerCase();
      if (i < n) i++;
      if (val) {
        if (neg) branch.negTerms.push(val);
        else branch.terms.push(val);
      }
      continue;
    }
    const start = i;
    while (i < n && !/\s/.test(s[i])) i++;
    const val = s.slice(start, i).toLowerCase();
    if (val) {
      if (neg) branch.negTerms.push(val);
      else branch.terms.push(val);
    }
  }
  return branch;
}

function branchMatches(p, b) {
  const h = haystack(p);
  for (const t of b.terms) if (!h.includes(t)) return false;
  for (const [f, needle] of b.fields) if (!fieldText(p, f).includes(needle)) return false;
  for (const t of b.negTerms) if (h.includes(t)) return false;
  for (const [f, needle] of b.negFields) if (fieldText(p, f).includes(needle)) return false;
  return true;
}

function snippetAround(content, idx, len, radius = 48) {
  const start = Math.max(0, idx - radius);
  const end = Math.min(content.length, idx + len + radius);
  const frag = content.slice(start, end).replace(/\n/g, ' ');
  return `${start > 0 ? '...' : ''}${frag}${end < content.length ? '...' : ''}`;
}

function buildSnippet(p, b) {
  const content = p.content || '';
  if (!content) return '';
  const cl = content.toLowerCase();
  for (const [f, needle] of b.fields) {
    if (f === 'content' && needle && cl.includes(needle)) {
      return snippetAround(content, cl.indexOf(needle), needle.length);
    }
  }
  for (const t of b.terms) {
    if (t && cl.includes(t)) return snippetAround(content, cl.indexOf(t), t.length);
  }
  for (const [f, needle] of b.fields) {
    if (f !== 'content' && needle && cl.includes(needle)) {
      return snippetAround(content, cl.indexOf(needle), needle.length);
    }
  }
  return '';
}

function matchFlags(p, b) {
  const titleL = String(p.title || '').toLowerCase();
  const cl = String(p.content || '').toLowerCase();
  const folderL = String(p.folder_path || '').toLowerCase();
  const deckL = String(p.deck_name || '').toLowerCase();
  const tagsL = (p.tags || []).map((t) => String(t || '').toLowerCase()).join(' ');

  const titleMatch =
    b.terms.some((t) => titleL.includes(t)) ||
    b.fields.some(([f, n]) => f === 'title' && titleL.includes(n));
  const contentMatch =
    b.terms.some((t) => cl.includes(t)) || b.fields.some(([f, n]) => f === 'content' && cl.includes(n));
  const folderMatch =
    b.terms.some((t) => folderL.includes(t)) ||
    b.fields.some(([f, n]) => f === 'folder' && folderL.includes(n));
  const deckMatch =
    b.terms.some((t) => deckL.includes(t)) || b.fields.some(([f, n]) => f === 'deck' && deckL.includes(n));
  const tagMatch =
    b.terms.some((t) => tagsL.includes(t)) || b.fields.some(([f, n]) => f === 'tag' && tagsL.includes(n));

  return { title_match: titleMatch, content_match: contentMatch, folder_match: folderMatch, deck_match: deckMatch, tag_match: tagMatch };
}

export function searchPapersAdvanced(papers, query) {
  const q = String(query || '').trim();
  if (!q) return [];
  const branches = splitOrBranches(q).map(parseBranch);
  const out = [];
  for (const p of papers) {
    let matched = null;
    for (const br of branches) {
      if (branchMatches(p, br)) {
        matched = br;
        break;
      }
    }
    if (!matched) continue;
    const flags = matchFlags(p, matched);
    out.push({
      id: p.id,
      title: p.title,
      folder_path: p.folder_path,
      deck_name: p.deck_name || '',
      snippet: buildSnippet(p, matched),
      ...flags,
    });
  }
  return out;
}
